compiler/src/iree/compiler/Dialect/HAL/IR/HALAttrs.cpp

```diff
@@ -1039,6 +1039,89 @@ bool DevicePromiseAttr::isLegalToInline(Operation *inlineSite,
   return *this == targetAffinityAttr;
 }
 
+//===----------------------------------------------------------------------===//
+// #hal.device.topology<...>
+//===----------------------------------------------------------------------===//
+
+bool DeviceTopologyAttr::requiresTransfer(
+    IREE::Stream::AffinityAttr source,
+    IREE::Stream::AffinityAttr target) const {
+
+  auto sourceDevice = mlir::dyn_cast_if_present<DeviceAffinityAttr>(source);
+  auto targetDevice = mlir::dyn_cast_if_present<DeviceAffinityAttr>(target);
+
+  if (!sourceDevice || !targetDevice)
+    return true;
+  if (sourceDevice == targetDevice)
+    return false;
+
+  SymbolRefAttr sourceRef = sourceDevice.getDevice();
+  SymbolRefAttr targetRef = targetDevice.getDevice();
+
+  // Check if our topology has a link between the two devices
```

Comment:
style: proper punctuation in comments (here and in the other lines below)
```suggestion
  // Check if our topology has a link between the two devices.
```